src/parameterizations/vertical/MOM_geothermal.F90

```diff
@@ -304,6 +316,14 @@ subroutine geothermal(h, tv, dt, ea, eb, G, GV, CS, halo)
     enddo ; endif
   enddo ! j-loop
 
+! Calculate heat tendency due to addition and transfer of internal heat
+if (CS%id_internal_heat_tend_3d > 0) then
+  do k=1,nz ; do j=js,je ; do i=is,ie
+    work_3d(i,j,k) = GV%H_to_kg_m2 * tv%C_p * Idt * (h(i,j,k) * tv%T(i,j,k) - h_old(i,j,k) * temp_old(i,j,k))
```

Comment:
I think we may be OK here, since the multiply pairs will happen prior to the addition, and multiplication order is not generally an issue.  But good to get feedback from grandmaster @Hallberg-NOAA if possible.  (He's currently chasing dragons in Indonesia...)

Reply:
You'll have to be more specific about what you mean, Fortran is not my forté! You mean parentheses around the other variables in the product?

Reply:
Yes, so `a*b*c = a*(b*c) = (a*b)*c` in full precision, but not in finite precision. @Hallberg-NOAA wagged a finger at me for not using enough parentheses in a few of my routines.

Reply:
Just trying to save you a dragon-backed finger wag ;)

Reply:
I still think this one might be OK. For example, no parentheses here in `MOM_hor_visc.F90`:
```
 520     do j=Jsq-1,Jeq+2 ; do i=Isq-1,Ieq+2                                                                                                                                                  
 521       dudx(i,j) = CS%DY_dxT(i,j)*(G%IdyCu(I,j) * u(I,j,k) - &                   
 522                                   G%IdyCu(I-1,j) * u(I-1,j,k))                  
 523       dvdy(i,j) = CS%DX_dyT(i,j)*(G%IdxCv(i,J) * v(i,J,k) - &                   
 524                                   G%IdxCv(i,J-1) * v(i,J-1,k))                  
 525       sh_xx(i,j) = dudx(i,j) - dvdy(i,j)                                        
 526     enddo ; enddo                                                               
```
AFAIK the compiler will treat `a*b + c*d` as `(a*b) + (c*d)`.

Reply:
So is it a simple case of adding parentheses for each individual operation, e.g. `a*(b*(c-d))` or `(a*b)*(c-d)` rather than `a*b*(c-d)`? Or does the positioning of the parentheses require more detailed thought?

Reply:
Yeah it's not an order of operations issue. In the example that you showed, it's unambiguous what the order of multiplications are because `G%IdxCv(i,J) * v(i,J,k) - G%IdxCv(i,J-1) * v(i,J-1,k)` gets evaluated first. It's more that the commutative property of multiplication doesn't hold in finite precision such that (using the @gmacgilchrist's code above ignoring the change in heat  term)
`GV%H_to_kg_m2 * tv%C_p * Idt` could be evaluated in six different ways (right to left, left to right, and other permutations) which would all be equivalent for infinite precision, but might all differ at the roundoff level. There's an example of this when comparing `tracer_vertdiff`vs. `triDiagTS`.

Reply:
@gmacgilchrist, if you really want to go far down the rabbit-hole (which I don't recommend), you could probably try to group the parentheses such that as much precision as possible is preserved. It's unnecessary for something like this; just sprinkle enough in to make the order unambiguous.

Reply:
@ashao I very much agree with you that multiplication is not always associative, and should have addressed your original example of `a*b*c`.  From my discussions with @Hallberg-NOAA, I had the impression that associativity of addition is strictly enforced since they come up often (e.g. `1 + 1e-40 - 1e-40 = ??`) but that multiplicative associativity was not considered a severe problem since the non-associative cases are uncommon, and therefore not (yet) strictly enforced.  That's why I wasn't going to raise the issue.

But if I look at the source, I can see that Bob is actually starting to enforce associativity of multiplication.  For example, [this block](https://github.com/NOAA-GFDL/MOM6/blob/c7216cc1e541dfa88554d153461d5d5c215a4316/src/parameterizations/vertical/MOM_diabatic_driver.F90#L1010) in `MOM_diabatic_driver.F90` from two months ago touched by Bob does not use parentheses:
```
      do k=2,nz ; do j=js,je ; do i=is,ie
        hval = 1.0 / (h_neglect + 0.5*(h(i,j,k-1) + h(i,j,k)))
        ea_t(i,j,k) = (GV%Z_to_H**2) * dt_in_T * hval * Kd_heat(i,j,k)
        eb_t(i,j,k-1) = ea_t(i,j,k)
        ea_s(i,j,k) = (GV%Z_to_H**2) * dt_in_T * hval * Kd_salt(i,j,k)
        eb_s(i,j,k-1) = ea_s(i,j,k)
      enddo ; enddo ; enddo
```
but [this block](https://github.com/NOAA-GFDL/MOM6/blob/c7216cc1e541dfa88554d153461d5d5c215a4316/src/parameterizations/vertical/MOM_diabatic_driver.F90#L1123) from one day later does enforce multiplicative associativity:
```
          if (CS%use_legacy_diabatic) then
            add_ent = ((dt_in_T * visc%Kd_extra_S(i,j,k)) * GV%Z_to_H**2) / &
               (0.25 * ((h(i,j,k-1) + h(i,j,k)) + (hold(i,j,k-1) + hold(i,j,k))) +  h_neglect)
          else
            add_ent = ((dt_in_T * visc%Kd_extra_S(i,j,k)) * GV%Z_to_H**2) / &
               (0.5 * (h(i,j,k-1) + h(i,j,k)) + &
                h_neglect)
```
So, I think I have to concede here that you are right, and that it would be better to enforce the order if possible.

Sorry @gmacgilchrist, this has turned into a very wonky discussion.  But hopefully these examples will at least guide you a bit.

Reply:
Yep fair enough, thanks for sharing that information with me @marshallward! Also thanks to @gmacgilchrist for allowing us to highjack your PR review.

Reply:
No problem, always happy to be learning!

So, in this particular case, something like:
`((GV%H_to_kg_m2 * tv%C_p) * Idt )* (h(i,j,k) * tv%T(i,j,k) - h_old(i,j,k) * temp_old(i,j,k))`
would be preferred?

And the reason is that it removes ambiguity with regards the order of multiplication?

Reply:
Exactly. I suppose as @marshallward notes this isn't explicitly put in the style guide for MOM6 but  maybe something that may come down the pipeline.

Reply:
Just had a chat with @adcroft, he confirmed that multiplication order has not traditionally been enforced.  But after our convo here and seeing the examples in MOM6 Slack, I'm at least convinced that @ashao is right and this is something we need to start addressing.

For this PR I'd say it's up to you Graeme, but definitely something to consider going forward.